Adding documentation and tests for the ColumnTypes type class.

imarios opened the pull request.

imarios commented:
Hey @kanterov this is a pretty benign change. Adds a bit more documentation and tests. Let me know if you have any comments. thanks!

kanterov commented on dataset/src/main/scala/frameless/ops/ColumnTypes.scala:
```diff
@@ -3,17 +3,25 @@ package ops
-    implicit tail: ColumnTypes.Aux[V, TT, T]
-  ): ColumnTypes.Aux[V, TypedColumn[V, H] :: TT, H :: T] =
-    new ColumnTypes[V, TypedColumn[V, H] :: TT] {type Out = H :: T}
+  implicit def deriveCons[T, Head, Tail <: HList, ColTypes <: HList](
```
What is the reasoning behind renaming? It falls off current coding style, as well as other OSS projects using shapeless

imarios replied:
Hey @kanterov, you mean about the head and tail? Yeah, that one I guess makes sense to use the standard H and TT. Overall the main goal of the  rename was to make it a bit easier to understand the code. Maybe we can revert the Head/Tail back to their standard notations and leave the ColTypes as is (since it has a well defined semantic in our case)?

kanterov commented on dataset/src/main/scala/frameless/ops/ColumnTypes.scala:
```diff
@@ -3,17 +3,25 @@ package ops
+  *   type Out = A :: B :: C :: HNil
+  * }}}
+  */
+trait ColumnTypes[T, U <: HList] {
```
This is mostly a workaround to issue with slow implicit derivation for `Comapped`

imarios replied:
Oh, just realized that this is a special case of using `Comapped`! Ok, I think I will add this comment in case someone in the future decide to revert this back to Comapped.

imarios commented:
@kanterov I just unified the notation for ColumnsTypes and AggregateType (they were similar implementations but used different letters for their equivalent parameters). I reverted the notation back to the standard ones used by shapeless (as you suggested). Added some comments as well. Let me know if there is anything else. I will squash the three commits into one after you give a thumps up.

imarios commented:
@kanterov hello! anything else to improve here?
